### ISSUE TITLE
Problem in solving a SDP written in JuMP

### ISSUE THREAD

**Issue by ganglesmunne** [nanleij/ClusteredLowRankSolver.jl#46]
Hello,

I am trying to use ClusteredLowRankSolver to solve a SDP written in JuMP.

For doing this, I am using the function from JuMp to create a file in .dat-s 

> write_to_file(model, "pb.dat-s")

and then I use

> problem=sdpa_sparse_to_problem("pb.dat-s")

to obtain the SDP  in the ClusteredLowRankSolver format as shown in https://docs.juliahub.com/ClusteredLowRankSolver/VwoDC/1.0.9/sdpa/ .

This process does not give me any error. However, when I tried to solve the problem using

> status, primalsol, dualsol, time, errorcode=solvesdp(problem)

it reports the following error:

> ArgumentError: collection must be non-empty
> 
> Stacktrace:
>  [1] first(itr::Base.KeySet{Any, Dict{Any, Any}})
>    @ Base ./abstractarray.jl:473
>  [2] ClusteredLowRankSDP(sos::Problem; prec::Int64, verbose::Bool)
>    @ ClusteredLowRankSolver ~/.julia/packages/ClusteredLowRankSolver/v5JhL/src/interface.jl:874
>  [3] ClusteredLowRankSDP
>    @ ~/.julia/packages/ClusteredLowRankSolver/v5JhL/src/interface.jl:836 [inlined]
>  [4] #solvesdp#318
>    @ ~/.julia/packages/ClusteredLowRankSolver/v5JhL/src/solver.jl:62 [inlined]
>  [5] solvesdp(problem::Problem)
>    @ ClusteredLowRankSolver ~/.julia/packages/ClusteredLowRankSolver/v5JhL/src/solver.jl:37
>  [6] top-level scope
>    @ In[10]:6

Do you know if something might be wrong in the conversion from JuMp to SDPA and from SDPA to ClusteredLowRankSolver?

**Comment by nanleij** [comment]
Hi, 

This looks like there is a constraint without positive semidefinite variables. Due to the conversion to SDPA format and back this might not directly translate to a constraint on your side; it seems like JuMP transforms the SDPs to the primal format (see [SDPA-sparse format](https://plato.asu.edu/ftp/sdpa_format.txt)), and then writes that to the file, instead of using the dual format which is used in ClusteredLowRankSolver.

Running an [example](https://jump.dev/JuMP.jl/stable/tutorials/conic/simple_examples/#The-correlation-problem) from JuMP (using that code to build the model, then write it to a ,dat-s file and reading with sdpa_sparse_to_problem) works for me, so I don't know exactly where it goes wrong. 

It might be that you use a matrix variable without declaring it to be PSD. I might be able to say more if you can make a minimum working example.

**Comment by ganglesmunne** [comment]
Hi, thank you very much for your insights!

My SDP is a bit complicate, but here I attached a small example written in JuMP,

![imagen](https://github.com/user-attachments/assets/01339a84-3f92-4eca-aa06-282331e4fe38)

where the real variables 

> ymatrix0, ymatrix1, ymatrix2, ymatrix3, A0, P0 .

This example is enough to have the error mentioned in the previous message.

**Comment by ganglesmunne** [comment]
Here I also attached how the .dat-s converted with the juMp function looks like:

8
7
-1 -1 -1 -1 -1 2 1
1.0 -0.0 -0.0 -0.0 0.5 -0.0 -1.0 -0.0
1 1 1 1 2.0
5 1 1 1 0.5
7 1 1 1 1.0
1 2 1 1 -2.0
5 2 1 1 -0.5
7 2 1 1 -1.0
2 3 1 1 1.1547005383792515
3 3 1 1 0.3333333333333333
4 3 1 1 0.3333333333333333
5 3 1 1 -0.5
7 3 1 1 1.0
2 4 1 1 -1.1547005383792515
3 4 1 1 -0.3333333333333333
4 4 1 1 -0.3333333333333333
5 4 1 1 0.5
7 4 1 1 -1.0
0 5 1 1 -1.0
1 5 1 1 1.0
5 5 1 1 0.5
7 5 1 1 -1.0
1 6 1 1 1.0
2 6 1 2 1.0
3 6 2 2 1.0
4 7 1 1 1.0

**Comment by nanleij** [comment]
In the new version, it should work. It is probably still a good idea to check your code to construct the SDP in JuMP, since you're apparently not using a variable that you define, so it might indicate that you're not solving the SDP you want to solve.

**Comment by ganglesmunne** [comment]
Hi, thank you very much for the quick reply and to update the package, it was really helpful. 

Indeed, the problem was some variables I was not using. 

However, I am facing another problem when I try to solve the same problem with the ClusteredLowRankSolver as shown in the following picture:

![imagen](https://github.com/user-attachments/assets/9ea51f8f-43a7-4a53-87c0-98c8bc6146e9)

However, with other solvers I obtain a feasible solution.

**Comment by nanleij** [comment]
I think what is happening now is that there are linearly dependent constraints, which translate (I think) in your initial formulation to linearly dependent real variables (in your example SDP, A0 and P0 are linearly dependent). I haven't gotten around to implementing preprocessing steps or handling this during the solving, so currently the only way to deal with this is to make sure you don't have linear dependencies in the program.

**Comment by ganglesmunne** [comment]
I see... but then any equality constraint relating two variables, would lead to linearly dependent real variables right?

For example, given two variables A and B, 
the constraint aA+bB=0 with a,b being real numbers, 
gives A=(b/a) B.

**Comment by nanleij** [comment]
That's right with 1 constraint, but often you have more constraints. For example, in the system of equations
aA + bB = 0
cA + dB = 0
the variables A and B are only linearly dependent if b/a = d/c.

**Comment by nanleij** [comment]
I just learned of the package Dualization.jl, which might help a bit. By default, jump writes the model in the primal (or geometric) form, whereas ClusteredLowRankSolver interprets it in the dual (standard) form (with equality constraints like you have). With Dualization.jl, you can use
```
write_to_file(dualize(model), filename)
```
so that both JuMP and ClusteredLowRankSolver use the same problem. 

This will not help with the linear dependencies, but interpreting the solution will be much easier. SDPA format is not designed for free variables, so those might still give some trouble since they're modeled as x1-x2 with x1, x2 >= 0.  That can result in large entries in the solution.

**Comment by ganglesmunne** [comment]
Thank you very much for you explanations! now I understand which are the main problems. 

You are right, I haven't thought in how to read the free variables in SDPA format, that also will be an issue at some point.
